File: AIProjects/day_3/leetcode_3066.py
class Solution:
    def minOperations(self, nums: List[int], k: int) -> int:
        heapq.heapify(nums)  # Convert list into a min-heap (O(N))
        num_ops = 0

        while nums[0] < k:  # Keep processing until smallest element >= k
            if len(nums) < 2:  # Edge case: if fewer than 2 elements remain
                return -1

            # Extract the two smallest elements (O(log N) each)
            x = heapq.heappop(nums)
            y = heapq.heappop(nums)

            # Combine them with the given formula and push back to heap (O(log N))
            heapq.heappush(nums, min(x, y) * 2 + max(x, y))
            num_ops += 1

        return num_ops

File: AIProjects/day_3/leetcode_3066.cpp
class Solution {
public:
    int minOperations(vector<int>& nums, int k) {
        
    priority_queue<long, vector<long>, greater<long>> minHeap(nums.begin(), nums.end());  // Min-heap
    int num_ops = 0;

    while (true) {
        long x = minHeap.top();  // Smallest element
        if (x >= k) return num_ops;  // Stop condition
        minHeap.pop();  // Remove smallest

        //if (minHeap.empty()) return -1;  // Edge case: Not enough elements left

        long y = minHeap.top();  // Second smallest element
        minHeap.pop();  // Remove second smallest

        // Combine and push new element back
        minHeap.push(min(x, y) * 2 + max(x, y));
        num_ops++;
    }


    }
};
